### PR TITLE
BHoM_Engine: Add and Query Fragments

### LINKED ISSUE #1184: Physical_Engine: Modify AddFragments
#### Description:

As per our [discussion](https://github.com/BHoM/BHoM/issues/553), I am creating a method to `AddFragments` to PhysicalElements

### PR DESCRIPTION
**Pull request by michaldengusiak**
### Issues addressed by this PR

Closes #1184 

Add Modify AddFragments

### Test files

N/A

### Changelog

- Add node to Modify `AddFragments`
- Add node Query `FindFragment`

### CODE REVIEW

**Comment by michaldengusiak** [comment]
ok, I implemented comments and moved to the correct place as advice

File: BHoM_Engine/Modify/AddFragment.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace BH.Engine.Base
+{
+    public static partial class Modify
+    {
+        [Description("Appends a Fragment Property to a given BHoM Object")]
+        [Input("iBHoMObject", "Any object implementing the IBHoMObject interface that can have fragment properties appended to it")]
+        [Input("fragment", "Any fragment object implementing the IBHoMFragment interface to append to the object")]
+        [Output("iBHoMObject", "The BHoM object with the added fragment")]
+        public static IBHoMObject AddFragment(this IBHoMObject iBHoMObject, IBHoMFragment fragment)
```
Comment (FraserGreenroyd) [inline BHoM_Engine/Modify/AddFragment.cs]: As per the discussion on #1184 I would like to see an optional `bool` (set to `false` initially) that allows users to replace fragments if they already exist.
Similarly, if the input is `false`, we should check if the fragment exists and return an error if it does. See discussion on #1184 for more info.

Comment (IsakNaslundBh) [reply BHoM_Engine/Modify/AddFragment.cs]: Another thing on this method, You will need to clone the object to be changed as well as its list before modifying it. Either use the DeepClone method in the BHoMEngine, or shallow clone the object, and copy the fragment list. Right now, the incoming objects will be changed, changing things upstream in an visual programming workflow.

File: BHoM_Engine/Modify/AddFragment.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        [Output("iBHoMObject", "The BHoM object with the added fragment")]
+        public static IBHoMObject AddFragment(this IBHoMObject iBHoMObject, IBHoMFragment fragment, bool replace = false)
+        {
+            if (iBHoMObject == null) return null;
+            IBHoMObject o = iBHoMObject.DeepClone();
+            o.Fragments = new List<IBHoMFragment>(iBHoMObject.Fragments);
+            if(o.Fragments.Contains(fragment))
+            {
+                if(replace)
+                {
```
Comment (FraserGreenroyd) [inline BHoM_Engine/Modify/AddFragment.cs]: Use single line `if` statements given these are single line operations

File: BHoM_Engine/Modify/AddFragment.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                    o.Fragments[o.Fragments.IndexOf(fragment)] = fragment;
+                }
+                else
+                {
+                    Reflection.Compute.RecordError("Fragment already exists");
```
Comment (FraserGreenroyd) [inline BHoM_Engine/Modify/AddFragment.cs]: Can we expand this error message to tell the user something like "That fragment already exists on this object. If you would like to replace the existing fragment set the 'replace' input to 'true'"

File: BHoM_Engine/Modify/AddFragment.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+using System.ComponentModel;
+
+namespace BH.Engine.Base
+{
+    public static partial class Modify
+    {
+        [Description("Appends a Fragment Property to a given BHoM Object")]
+        [Input("iBHoMObject", "Any object implementing the IBHoMObject interface that can have fragment properties appended to it")]
+        [Input("fragment", "Any fragment object implementing the IBHoMFragment interface to append to the object")]
+        [Input("replace", "If true will replace exisiting fragments ")]
```
Comment (FraserGreenroyd) [inline BHoM_Engine/Modify/AddFragment.cs]: Expand the description to be more informative for the user, Something like "If set to true and the object already contains a fragment of the type being added, the fragment will be replaced by this instance"

File: BHoM_Engine/Modify/AddFragment.cs (hunk 2 above)
Comment (michaldengusiak) [reply BHoM_Engine/Modify/AddFragment.cs]: done

File: BHoM_Engine/Modify/AddFragment.cs (hunk 3 above)
Comment (michaldengusiak) [reply BHoM_Engine/Modify/AddFragment.cs]: done

File: BHoM_Engine/Modify/AddFragment.cs (hunk 4 above)
Comment (michaldengusiak) [reply BHoM_Engine/Modify/AddFragment.cs]: done

**Comment by FraserGreenroyd** [comment]
While this LGTM now and will be merged in to unblock on the workflows, @michaldengusiak please can you raise issues to deprecate the existing methods within the Environment Engine which do similar functionality. Please assign to me and put on the 2.4RC milestone. Cheers!

**Pull request merged**